### DISCUSSION TITLE
Does Prisma need internet connection when generating client from db?

### DISCUSSION THREAD

**Discussion by BeGj** [prisma/prisma#9228] [Q&A] [answered]
I ran `npm i prisma` and  `prisma init` while online, but later i were offline and wanted to generate client from my local postgres docker instance.
**While offline:** 
So i ran `prisma db pull` and it worked alright and the prisma scheme were generated... but then when i wanted to generate the client with `prisma generate` i recieved errors. As if Prisma tried to connect online?
Can anyone explain?
I did this in a fresh nestjs project.

**Error:**
It's requesting url `......../windows/query.engine.dll.node.gz.sha256`.

**Comment by ryands17** [comment, marked as answer]
@BeGj 👋 

The very first time, `prisma generate` needs the internet to download the binary for your Operating System. After this is done, subsequent `prisma generate` commands don't need the internet to work.

**Reply by BeGj** [reply]
You mean the files in node_modules/@prisma/engine? I thought i had them downloaded by doing `npm i prisma`

**Reply by ryands17** [reply]
Not on `prisma` but on installing `@prisma/client` as shown here in the docs: https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/install-prisma-client-typescript-postgres

**Comment by janpio** [comment]
Can you show the full error message you are getting? There might be some unexpected things going on.

**Reply by rankun203** [reply]
@janpio so we have set `PRISMA_QUERY_ENGINE_LIBRARY=custom/my-query-engine-unix`, and when run prisma client, it's still trying to access internet for the sha256 file

> Error: request to https://binaries.prisma.sh/all_commits/dc520b92b1ebb2d28dc3161f9f82e875bd35d727/debian-openssl-1.1.x/libquery_engine.so.node.gz.sha256 failed, reason: connect ETIMEDOUT 13.227.26.37:443\nerror Command failed with exit code 1.

Our server has pretty strict outbound traffic control, can we skip the sha256 verification process?

**Reply by rankun203** [reply]
Update, if we provide `PRISMA_QUERY_ENGINE_LIBRARY=custom/my-query-engine-unix`, it will not go fetch the sha256 file. Above was a mistake that the env was actually missing in that Docker build stage.

**Reply by janpio** [reply]
We are aware the download process is not optimal and will hopefully soon be able to rewrite good parts of it to make it simpler and more stable, with less side effects like download of checksum files.

**Reply by bmajkut** [reply]
Hello
could You please elaborate a little more how to prevent prisma from downloading additional data like .sha256 files? 
I am trying to run `prisma generate` function during docker container building and I am getting exactly same error.
It is really strange because I am able to download npm packages via `npm install` but I cannot download those `sha256` files.

**Reply by Secret1007** [reply]
> Update, if we provide `PRISMA_QUERY_ENGINE_LIBRARY=custom/my-query-engine-unix`, it will not go fetch the sha256 file. Above was a mistake that the env was actually missing in that Docker build stage.

could u show the details

**Comment by wvhulle** [comment]
Today I was without internet connection for a while and generating the client failed. Can generation be done without internet connection?

**Comment by dmpvost** [comment]
Same for me, I need to install on a secure network without internet
How to handle that?
How detect & install dependencies before ?

**Comment by jan4984** [comment]
can the downloading be moved to `npm install` ?

**Comment by Cat7373** [comment]
+1

**Reply by exotexot** [reply]
That would be also needed. As i wanna use prisma in an offline docker environment.

**Reply by dmpvost** [reply]
I was able to do it on environnement without internet

my issue was that the client and cli was not on the same version that why it's downloading

after the prisma generate work fine

**Comment by exotexot** [comment]
I was using Prisma in a dockerised NodeJS application.
At first it wasn't working for me, because my development platform differed from my deployment platform.
After forcing platform: linux/amd64 when building the images, it worked on my offline machine.

I have RUN npx prisma migrate in my Dockerfile.
